File: Champions/Annie/Q.lua
Vector2 = require 'Vector2' -- include 2d vector lib 

function onStartCasting()   

end

function onFinishCasting()
    addProjectileTarget("Disintegrate", castTarget, false)
end

function applyEffects()
    local AP = owner:GetStats().AbilityPower.Total*0.8
    local damage = 45 + (spellLevel*35) + AP

    if ( not ( castTarget == 0 ) ) and ( not isDead( castTarget ) ) then
        dealMagicalDamage(damage)
        if isDead(castTarget) then
            -- Mana recover
            lowerCooldown(0, totalCooldown / 2)
            local manaToRecover = 55 + (spellLevel*5)
            local newMana = owner:GetStats().CurrentMana + manaToRecover
            local maxMana = owner:GetStats().ManaPoints.Total
            if newMana >= maxMana then
                owner:GetStats().CurrentMana = maxMana
            else
                owner:GetStats().CurrentMana = newMana
            end
        end
    end
    destroyProjectile()
end

function onUpdate(diff)

end

File: Champions/Annie/Q.cs
using System;
using LeagueSandbox.GameServer.Logic.GameObjects;
using LeagueSandbox.GameServer.Logic.API;
using LeagueSandbox.GameServer.Logic.GameObjects.AttackableUnits;
using LeagueSandbox.GameServer.Logic.Scripting.CSharp;

namespace Spells
{
    public class Disintegrate : GameScript
    {
        public void OnActivate(Champion owner)
        {
        }

        public void OnDeactivate(Champion owner)
        {
        }

        public void OnStartCasting(Champion owner, Spell spell, AttackableUnit target)
        {
        }

        public void OnFinishCasting(Champion owner, Spell spell, AttackableUnit target)
        {
            spell.AddProjectileTarget("Disintegrate", target, false);
        }

        public void ApplyEffects(Champion owner, AttackableUnit target, Spell spell, Projectile projectile)
        {
            var ap = owner.Stats.TotalAbilityPower * 0.8f;
            var damage = 45 + spell.Level * 35 + ap;
            if (target != null && !ApiFunctionManager.IsDead(target))
            {
                target.TakeDamage(owner, damage, DamageType.DAMAGE_TYPE_MAGICAL, DamageSource.DAMAGE_SOURCE_SPELL,
                    false);
                if (target.IsDead)
                {
                    spell.LowerCooldown(0, spell.getCooldown());
                    float manaToRecover = 55 + spell.Level * 5;
                    var newMana = owner.Stats.CurrentPar + manaToRecover;
                    var maxMana = owner.Stats.TotalPar;
                    owner.Stats.CurrentPar = Math.Min(maxMana, newMana);
                }
            }

            projectile.setToRemove();
        }

        public void OnUpdate(double diff)
        {
        }
    }
}
